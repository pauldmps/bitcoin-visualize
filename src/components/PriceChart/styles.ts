import styled from "styled-components";

export const Wrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-top: 12px;
  width: 50%;
  margin-left: 25%;

  @media (max-width: 1024px) {
    width: 60%;
    margin-left: 20%;
  }
`;