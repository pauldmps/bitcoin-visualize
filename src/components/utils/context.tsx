import React, { useState } from 'react';

const currencies = ['USD', 'EUR', 'CNY', 'JPY', 'PLN'];
const defaultCurrency = currencies[0];

interface IContext {
  currencies: Array<string>;
  defaultCurrency: string;
  selectedCurrency: string;
  setSelectedCurrency: (currency: string) => void;
}

export const AppContext = React.createContext<IContext | undefined>(undefined);
export const AppContextProvider = ({
  children,
}: {
  children: React.ReactNode;
}) => {
  const [selectedCurrency, setSelectedCurrency] = useState(defaultCurrency);

  return (
    <AppContext.Provider
      value={{
        currencies,
        defaultCurrency,
        selectedCurrency,
        setSelectedCurrency,
      }}
    >
      {children}
    </AppContext.Provider>
  );
};