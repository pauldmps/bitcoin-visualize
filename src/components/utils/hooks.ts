import { useState, useEffect, useContext } from 'react';
import { AppContext } from './context';

const PRICE_API_URL = (currency: string) =>
  `https://api.coindesk.com/v1/bpi/currentprice/${currency}.json`;
const PRICE_HISTORY_API_URL = (currency: string) =>
  `https://api.coindesk.com/v1/bpi/historical/close.json?currency=${currency}`;

const getApiData = async (url: string) => {
  const response = await fetch(url);
  const responseJson = await response.json();
  return responseJson;
};

export const useGetBitcoinPriceAndHistory = (currency: string) => {
  const [price, setPrice] = useState('');
  const [history, setHistory] = useState([]);
  const [error, setError] = useState(null);

  useEffect(() => {
    (async () => {
      try {
        const response = await Promise.all([
          getApiData(PRICE_API_URL(currency)),
          getApiData(PRICE_HISTORY_API_URL(currency)),
        ]);
        const {
          bpi: {
            [currency]: { rate },
          },
        } = response[0];
        setPrice(rate);

        const { bpi } = response[1];
        setHistory(bpi);
      } catch (error: any) {
        setError(error);
      }
    })();
  }, [currency]);

  return { error, price, history };
};

export const useAppContext = () => {
  const appContext = useContext(AppContext);
  if (appContext === undefined) {
    throw new Error('AppContext must be used inside AppContextProvider');
  }

  return appContext;
};
