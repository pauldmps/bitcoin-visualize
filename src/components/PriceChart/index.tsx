import Chart from 'chart.js/auto';
import { useEffect } from 'react';
import { useAppContext, useGetBitcoinPriceAndHistory } from '../utils/hooks';
import { Wrapper } from './styles';

const PriceChart = () => {
  const { selectedCurrency } = useAppContext();
  const { history } = useGetBitcoinPriceAndHistory(selectedCurrency);

  const data = {
    datasets: [
      {
        label: `Price 1 BTC to ${selectedCurrency}`,
        data: history,
        borderColor: 'white',
        backgroundColor: 'green',
        fill: 'start',
      },
    ],
  };

  useEffect(() => {
    const ctx = document.getElementById('priceChart')! as HTMLCanvasElement;
    const chart = new Chart(ctx, {
      type: 'line',
      data,
      options: {},
    });

    return () => chart.destroy && chart.destroy();
  }, [data]);

  return (
    <Wrapper>
      <canvas id="priceChart" width="400" height="400"></canvas>
    </Wrapper>
  );
};

export default PriceChart;
