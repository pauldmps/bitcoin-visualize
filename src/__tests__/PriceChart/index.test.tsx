import 'jest-canvas-mock';
import { render } from '@testing-library/react';
import PriceChart from '../../components/PriceChart';
import { mockHistoryApiData } from './__mocks__/mockData';
import { AppContextProvider } from '../../components/utils/context';

jest.mock('../../components/utils/hooks', () => ({
  ...jest.requireActual('../../components/utils/hooks'),
  useGetBitcoinPriceAndHistory: () => {
    return {
      history: mockHistoryApiData.bpi,
    };
  },
}));

jest.mock('chart.js/auto', () => ({
  __esModule: true,
  default: jest.fn(),
}));

const setup = () =>
  render(
    <AppContextProvider>
      <PriceChart />
    </AppContextProvider>
  );

describe('PriceChart component tests', () => {
  it('it renders correctly', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });
});
