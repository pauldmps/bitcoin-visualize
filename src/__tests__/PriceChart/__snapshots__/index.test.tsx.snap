// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PriceChart component tests it renders correctly 1`] = `
<div>
  <div
    class="sc-bdvvtL cYOTqU"
  >
    <canvas
      height="400"
      id="priceChart"
      width="400"
    />
  </div>
</div>
`;
